alternator/serialization.cc

```diff
@@ -272,6 +272,30 @@ big_decimal unwrap_number(const rjson::value& v, std::string_view diagnostic) {
     }
 }
 
+std::optional<big_decimal> unwrap_number(const rjson::value& v) {
```

Comment:
It might be confusing to future users that one flavor of the function throws, and the other returns an optional, but they share a name. Perhaps this one could be `unwrap_number_noexcept`? Normally I wouldn't care, but lately my coroutinization patch broke /next because of non-obvious throw semantics in a function and it's still fresh in my memory (:

Reply:
I don't like the word "noexcept" because people would rightly expect that a function called `something_noexcept` would be, well, `noexcept`, but this function can't be because it calls `json::print` and `big_decimal::big_decimal` which can probably allocate and throw if out of memory.

Come to think of it, perhaps the strange function that "deserves" a strange name is the existing function, which takes a string and uses it to build an exception. I could call it `unwrap_number_throw` or `unwrap_number_checked` or something... Would that make sense?

By the way, now that I look at this again, I see something surprising - unwrap_number allows the element inside the JSON to be an actual number, instead of a string we expect it to be. Do you rememeber if this is deliberate? Does DynamoDB allow that in the input JSON? I'll write a test to check...

Reply:
Unrelated to this series but I'll throw the fix in here too - unwrap_number should not allow numbers in the JSON. Only strings.

Reply:
> Come to think of it, perhaps the strange function that "deserves" a strange name is the existing function, which takes a string and uses it to build an exception. I could call it `unwrap_number_throw` or `unwrap_number_checked` or something... Would that make sense?

Another reason I prefer this direction is that other functions we have, like unwrap_set and unwrap_list, also return some sort of empty value instead of throwing and don't take an error string as parameter. Unfortunately all these unwrap_* functions aren't consistent on how they mark an error (e.g., unwrap_list returns a null pointer, my new unwrap_number returns a null optional), but at least they are consistent about returning, not throwing, on parse error.

Reply:
Could all it `try_unwrap_number()`.

Reply:
Ok, I'll use this try_unwrap_number() name.